src/test.rs

```diff
@@ -17,6 +18,118 @@ macro_rules! t {
     };
 }
 
+// `repo_test! will
+macro_rules! repo_test {
+    ($test_name:ident, ($($repo_type:ident),+), $test_body:expr) => {
+        paste::item! {
+            $(#[test]
+            fn [<$test_name _ $repo_type:snake>]() {
+                #[allow(unused_variables)]
+                let (td, repo) = $crate::test::repo_init2($crate::test::RepoType::$repo_type);
+                ($test_body)(&repo);
+            })+
+        }
+    }
+}
+
+pub struct TempDirs {
+    main: TempDir,
+    _rest: Vec<TempDir>,
+}
+
+impl Deref for TempDirs {
+    type Target = TempDir;
+
+    fn deref(&self) -> &Self::Target {
+        &self.main
+    }
+}
+
+pub fn repo_init_typical() -> (TempDirs, Repository) {
+    let (td, repo) = repo_init();
+    let tds = TempDirs {
+        main: td,
+        _rest: vec![],
+    };
+    (tds, repo)
+}
+
+pub fn repo_init_bare() -> (TempDirs, Repository) {
+    let td = TempDir::new().unwrap();
+    let repo = Repository::init_bare(td.path()).unwrap();
+    {
+        let mut config = repo.config().unwrap();
+        config.set_str("user.name", "name").unwrap();
+        config.set_str("user.email", "email").unwrap();
+        let mut index = repo.index().unwrap();
+        let id = index.write_tree().unwrap();
+
+        let tree = repo.find_tree(id).unwrap();
+        let sig = repo.signature().unwrap();
+        repo.commit(Some("HEAD"), &sig, &sig, "initial", &tree, &[])
+            .unwrap();
+    }
+    let tds = TempDirs {
+        main: td,
+        _rest: vec![],
+    };
+    (tds, repo)
+}
+
+pub fn repo_init_bare_worktree() -> (TempDirs, Repository) {
+    let td = TempDir::new().unwrap();
+    let repo = Repository::init_bare(td.path()).unwrap();
+    {
+        let mut config = repo.config().unwrap();
+        config.set_str("user.name", "name").unwrap();
+        config.set_str("user.email", "email").unwrap();
+        let mut index = repo.index().unwrap();
+        let id = index.write_tree().unwrap();
+
+        let tree = repo.find_tree(id).unwrap();
+        let sig = repo.signature().unwrap();
+        repo.commit(Some("HEAD"), &sig, &sig, "initial", &tree, &[])
+            .unwrap();
+    }
+    let worktree_td = TempDir::new().unwrap();
+    std::fs::remove_dir(worktree_td.path()).unwrap(); // worktree will fail if the directory exists
+    let worktree = repo.worktree("worktree", worktree_td.path(), None).unwrap();
+    let worktree_repo = worktree.to_repository().unwrap();
+
+    let tds = TempDirs {
+        main: worktree_td,
+        _rest: vec![td],
+    };
+    (tds, worktree_repo)
+}
+
+pub fn repo_init_typical_worktree() -> (TempDirs, Repository) {
```

Comment:
Since these functions seem related to worktree-testing, perhaps they could be scoped to the worktree tests submodule?